--- a/drivers/hid/i2c-hid/i2c-hid.c	2015-02-07 12:02:56.677392663 -0600
+++ b/drivers/hid/i2c-hid/i2c-hid.c	2015-02-07 12:03:23.397425048 -0600
@@ -370,7 +370,7 @@
 static void i2c_hid_get_input(struct i2c_hid *ihid)
 {
 	int ret, ret_size;
-	int size = ihid->bufsize;
+    int size = le16_to_cpu(ihid->hdesc.wMaxInputLength);
 
 	ret = i2c_master_recv(ihid->client, ihid->inbuf, size);
 	if (ret != size) {
